File: common/src/main/java/com/bcm/messenger/common/recipients/RecipientProvider.java
package com.bcm.messenger.common.recipients;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.bcm.messenger.common.core.Address;
import com.bcm.messenger.common.database.records.RecipientSettings;
import com.bcm.messenger.common.database.repositories.RecipientRepo;
import com.bcm.messenger.common.database.repositories.Repository;
import com.bcm.messenger.utility.dispatcher.AmeDispatcher;
import com.bcm.messenger.utility.AppContextHolder;

import com.bcm.messenger.utility.concurrent.ListenableFutureTask;
import com.bcm.messenger.utility.Util;

import java.util.List;
import java.util.Map;
import java.util.WeakHashMap;
import java.util.concurrent.ExecutorService;

import kotlin.Unit;
import kotlin.jvm.functions.Function0;
import com.bcm.messenger.utility.logger.ALog;

/**
 * 联系人提供器
 */
class RecipientProvider {

    private static final String TAG = RecipientProvider.class.getSimpleName();

    private static RecipientCache reserveCache = null;//后备缓存（所有，包括群组里的陌生人, 目前考虑全部联系人都用这种缓存)

    private static final ExecutorService asyncRecipientResolver = Util.newSingleThreadedLifoExecutor();

    public RecipientProvider() {
        reserveCache = new RecipientCache(new WeakHashMap<>(5000));
    }

    /**
     * 清理所有缓存
     */
    void clearCache() {
        reserveCache.reset();
    }


    /**
     * 从缓存读取联系人信息
     * @param address
     * @return
     */
    @Nullable Recipient findCache(@NonNull Address address) {
        if (address.isGroup()) {
            return reserveCache.get(address);
        }else {
            return reserveCache.get(address);
        }
    }

    /**
     * 更新缓存（主要几种情况，自身，陌生人，以及本身通讯录的联系人）
     * @param address
     * @param recipient
     */
    void updateCache(@NonNull Address address, @Nullable Recipient recipient) {
        Recipient r = recipient;
        if (address.isCurrentLogin() && null == r) {
            r = Recipient.from(AppContextHolder.APP_CONTEXT, address, true);
        }
        reserveCache.set(address, r);
    }

    /**
     * 检查是否使用缓存
     * @param cachedRecipient
     * @param asynchronous
     * @return
     */
    private boolean useCache(Recipient cachedRecipient, boolean asynchronous) {

        if (cachedRecipient == null) {
            ALog.d(TAG, "useCache fail, cacheRecipient is null");
            return false;
        } else if (cachedRecipient.isStale()) {
            return false;
        } else if (!asynchronous && cachedRecipient.isResolving()) {
            return false;
        }
//        else if (hasGroup && hasSettings && (cachedRecipient.isResolving() || !cachedRecipient.isGroupRecipient() && !cachedRecipient.isSystemContact())) {
//            return false;
//        }
        return true;

    }

    @NonNull Recipient getRecipient(Context context, Address address, @Nullable RecipientDetails details, boolean asynchronous) {
        RecipientDetails newDetail = details == null ? new RecipientDetails(null, null, null, null) : details;

        Recipient current = findCache(address);
        if (!useCache(current, asynchronous)) {//如果不用缓存的数据，则重新生成recipient，并从数据库加载setting（同步或异步由asynchronous来定）

            ListenableFutureTask<RecipientDetails> futureTask = null;
            if (asynchronous) {
                futureTask = getRecipientDetailsAsync(context, address, newDetail);
                current = new Recipient(address, current, newDetail, futureTask);
            } else {
                fetchRecipientDetails(context, address, newDetail, false);
                current = new Recipient(address, current, newDetail, null);
            }
            updateCache(address, current);


        } else {
            final Recipient target = current;
            if(target != null) {
                if (asynchronous) {
                    AmeDispatcher.INSTANCE.getIo().dispatch(new Function0<Unit>() {
                        @Override
                        public Unit invoke() {
                            target.updateRecipientDetails(details, true);
                            return null;
                        }
                    });
                }else {
                    target.updateRecipientDetails(details, true);
                }
            }
        }

        return current;
    }

    private @NonNull
    ListenableFutureTask<RecipientDetails> getRecipientDetailsAsync(final Context context, final @NonNull Address address, final @NonNull RecipientDetails detail) {

        ListenableFutureTask<RecipientDetails> future =
                new ListenableFutureTask<>(() -> {
                    fetchRecipientDetails(context, address, detail, true);
                    return detail;
                });
        asyncRecipientResolver.submit(future);
        return future;
    }

    private void fetchRecipientDetails(Context context, @NonNull Address address, @NonNull RecipientDetails details, boolean nestedAsynchronous) {
        if (address.isGroup()) {
            fetchGroupRecipientDetails(context, address, details, nestedAsynchronous, true);
        } else {
            fetchIndividualRecipientDetails(context, address, details, true);
        }
    }

    private void fetchIndividualRecipientDetails(Context context, @NonNull Address address, @NonNull RecipientDetails details, boolean force) {
        RecipientRepo recipientRepo = Repository.getRecipientRepo();
        if (details.getSettings() == null && force && recipientRepo != null) {
            details.setSettings(recipientRepo.getRecipient(address.serialize()));
        }

    }

    private void fetchGroupRecipientDetails(Context context, Address groupId, @NonNull RecipientDetails details, boolean asynchronous, boolean force) {

        if ((details.participants == null || details.participants.isEmpty()) && force) {
//            Optional<GroupRecord> groupRecord = DatabaseFactory.getGroupDatabase(context).getGroup(groupId.toGroupString());
            List<Recipient> members = null;
            String groupTitle = null;
//            if (groupRecord.isPresent()) {
//                groupTitle = groupRecord.get().getTitle();
//                List<Address> memberAddresses = groupRecord.get().getMembers();
//                members = new LinkedList<>();
//                for (Address memberAddress : memberAddresses) {
//                    members.add(getRecipient(context, memberAddress, details, asynchronous));
//                }
//                details.setCustomName(groupTitle);
//                details.setParticipants(members);
//            }
        }
        if (details.settings == null && force) {
            details.setSettings(Repository.getRecipientRepo().getRecipient(groupId.serialize()));
        }

    }

    /**
     * 联系人详情信息
     */
    public static class RecipientDetails {

        @Nullable
        String customName;//针对群组是群组名称，针对个人是手机通讯录名称
        @Nullable
        String customAvatar;//针对群组是群组头像，针对个人是手机通讯录名称
        @Nullable
        List<Recipient> participants;
        @Nullable
        RecipientSettings settings;

        RecipientDetails(@Nullable String name, @Nullable String avatar,
                         @Nullable RecipientSettings settings,
                         @Nullable List<Recipient> participants) {
            this.customName = name;
            this.customAvatar = avatar;
            this.participants = participants;
            this.settings = settings;

        }

        @Nullable
        public String getCustomName() {
            return customName;
        }

        public void setCustomName(@Nullable String customName) {
            this.customName = customName;
        }

        @Nullable
        public String getCustomAvatar() {
            return customAvatar;
        }

        public void setCustomAvatar(@Nullable String customAvatar) {
            this.customAvatar = customAvatar;
        }

        @Nullable
        public List<Recipient> getParticipants() {
            return participants;
        }

        public void setParticipants(@Nullable List<Recipient> participants) {
            this.participants = participants;
        }

        @Nullable
        public RecipientSettings getSettings() {
            return settings;
        }

        public void setSettings(@Nullable RecipientSettings settings) {
            this.settings = settings;
        }
    }

    /**
     * 联系人缓存信息
     */
    private static class RecipientCache {
        
        private Map<Address, Recipient> cache;

        RecipientCache(Map<Address, Recipient> cache) {
            this.cache = cache;
        }

        public synchronized @Nullable
        Recipient get(Address address) {
            return cache.get(address);
        }

        public synchronized void set(Address address, @Nullable Recipient recipient) {
            if (recipient != null) {
                cache.put(address, recipient);
            } else {

                Recipient temp = cache.get(address);
                if (temp != null) {
                    temp.setStale();
                }
                cache.remove(address);

            }
        }

        public synchronized void reset() {
            for (Recipient recipient : cache.values()) {
                recipient.setStale();
            }
            cache.clear();
        }

        synchronized void remove(Address address) {
            this.cache.remove(address);
        }

    }

}

File: common/src/main/java/com/bcm/messenger/common/recipients/RecipientProvider.kt
package com.bcm.messenger.common.recipients

import android.content.Context
import com.bcm.messenger.common.AccountContext
import com.bcm.messenger.common.core.Address
import com.bcm.messenger.common.database.records.RecipientSettings
import com.bcm.messenger.common.database.repositories.RecipientRepo
import com.bcm.messenger.common.database.repositories.Repository
import com.bcm.messenger.utility.AppContextHolder
import com.bcm.messenger.utility.LRUCache
import com.bcm.messenger.utility.dispatcher.AmeDispatcher
import com.bcm.messenger.utility.logger.ALog
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import java.util.*
import java.util.concurrent.TimeUnit


internal class RecipientProvider(private val mAccountContext: AccountContext) {

    companion object {

        private val TAG = "RecipientProvider"

        private var mCommonCache: RecipientCache = RecipientCache(WeakHashMap(3000))

        private var mUncommonCache: RecipientCache = RecipientCache(LRUCache(5000))

        @Synchronized
        fun clearCache() {
            mCommonCache.reset()
            mUncommonCache.reset()
        }

        @Synchronized
        fun findCache(address: Address): Recipient? {
            return mCommonCache[address] ?: mUncommonCache[address]
        }

        @Synchronized
        fun updateCache(recipient: Recipient) {
            val r = recipient
            if (r.isContextLogin || (r.relationship != RecipientRepo.Relationship.STRANGER && r.relationship != RecipientRepo.Relationship.REQUEST)) {
                mUncommonCache.remove(r.address)
                mCommonCache[r.address] = r
            }else {
                mCommonCache.remove(r.address)
                mUncommonCache[r.address] = r
            }
        }

        @Synchronized
        fun deleteCache(address: Address) {
            mCommonCache.remove(address)
            mUncommonCache.remove(address)
        }

        /**
         *
         * @param cachedRecipient
         * @param asynchronous
         * @return
         */
        private fun useCache(cachedRecipient: Recipient?, asynchronous: Boolean): Boolean {

            if (cachedRecipient == null) {
                ALog.d(TAG, "useCache fail, cacheRecipient is null")
                return false
            } else if (cachedRecipient.isStale) {
                return false
            } else if (!asynchronous && cachedRecipient.isResolving) {
                return false
            }
            return true

        }
    }

    private var mTaskCounter = false

    @Volatile
    private var mTaskDisposable: Disposable? = null

    private var mTargetList: MutableList<Recipient> = mutableListOf()

    @Synchronized
    fun getRecipient(context: Context, address: Address, details: RecipientDetails?, asynchronous: Boolean): Recipient {
        var current = findCache(address)
        if (!useCache(current, asynchronous)) {
            current = Recipient(address, current)
            if (asynchronous) {
                handleFetchDetailTask(context, current)
            } else {
                val newDetail = details ?: RecipientDetails(address.serialize(), null, null, null, null)
                updateRecipientDetails(context, current, newDetail)
            }

        } else {
            current?.updateRecipientDetails(details, true)
        }
        return current!!
    }

    private fun updateRecipientDetails(context: Context, recipient: Recipient, details: RecipientDetails) {
        if (recipient.isGroupRecipient) {
            updateGroupRecipientDetails(context, recipient.address, details)
        } else {
            updateIndividualRecipientDetails(context, recipient.address, details)
        }
        recipient.updateRecipientDetails(details, false)
    }

    private fun updateIndividualRecipientDetails(context: Context, address: Address, details: RecipientDetails) {
        val recipientRepo = Repository.getRecipientRepo(mAccountContext)
        if (recipientRepo != null && details.settings == null) {
            details.settings = recipientRepo.getRecipient(address.serialize())
        }
    }

    private fun getRecipientDetails(context: Context, addressMap: Map<String, Recipient>) {
        val recipientRepo = Repository.getRecipientRepo(mAccountContext)
        if (recipientRepo != null) {
            val settingList = recipientRepo.getRecipients(addressMap.map {
                ALog.d(TAG, "getRecipientDetails uid: ${it.key}")
                it.key
            })
            settingList.forEach {
                addressMap[it.uid]?.updateRecipientDetails(RecipientDetails(it.uid, null, null, it, null))
            }
        }
        for ((uid, r) in addressMap) {
            if (r.isResolving) {
                r.updateRecipientDetails(RecipientDetails(uid, null, null, null, null))
            }
        }
    }

    private fun updateGroupRecipientDetails(context: Context, groupId: Address, details: RecipientDetails) {
        val recipientRepo = Repository.getRecipientRepo(mAccountContext)
        if (recipientRepo != null && details.settings == null) {
            details.settings = recipientRepo.getRecipient(groupId.serialize())
        }
    }

    @Synchronized
    private fun lockTaskCounter(recipient: Recipient): Boolean {
        mTargetList.add(recipient)
        return if (!mTaskCounter) {
            mTaskCounter = true
            true
        }else {
            false
        }
    }

    @Synchronized
    private fun releaseTaskCounter(): Map<String, Recipient> {
        return if (mTaskCounter) {
            mTaskCounter = false
            val map = mutableMapOf<String, Recipient>()
            mTargetList.forEach {
                map[it.address.serialize()] = it
            }
            mTargetList.clear()
            map
        }else {
            mutableMapOf<String, Recipient>()
        }
    }

    /**
     * handle recipient init task
     */
    private fun handleFetchDetailTask(context: Context, recipient: Recipient) {
        ALog.d(TAG, "handleFetchDetailTask uid: ${recipient.address}")
        updateCache(recipient)
        if (lockTaskCounter(recipient)) {
            mTaskDisposable = Observable.create<Boolean> {
                ALog.d(TAG, "handleFetchDetailTask begin")
                val targetMap = releaseTaskCounter()
                getRecipientDetails(AppContextHolder.APP_CONTEXT, targetMap)
                it.onNext(true)
                it.onComplete()

            }.delaySubscription(500, TimeUnit.MILLISECONDS, AmeDispatcher.ioScheduler)
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe({
                        ALog.d(TAG, "handleFetchDetailTask end")
                    }, {
                        ALog.e(TAG, "handleFetchDetailTask error", it)
                    })

        }
    }


    class RecipientDetails internal constructor(var uid: String,
                                                var customName: String?,
                                                var customAvatar: String?,
                                                var settings: RecipientSettings?,
                                                var participants: List<Recipient>?)

    private class RecipientCache internal constructor(private val cache: MutableMap<Address, Recipient>) {

        @Synchronized
        operator fun get(address: Address): Recipient? {
            return cache[address]
        }

        @Synchronized
        operator fun set(address: Address, recipient: Recipient?) {
            if (recipient != null) {
                cache[address] = recipient
            } else {
                val temp = cache[address]
                temp?.setStale()
                cache.remove(address)

            }
        }

        @Synchronized
        fun reset() {
            for (recipient in cache.values) {
                recipient.setStale()
            }
            cache.clear()
        }

        @Synchronized
        internal fun remove(address: Address) {
            this.cache.remove(address)
        }

    }

}